### DISCUSSION TITLE
Feedback on the ZOSLIB library

### DISCUSSION THREAD

**Discussion by IgorTodorovskiIBM** [zopencommunity/meta#62] [General]
Hi, this discussion is aimed at collecting feedback for the ZOSLIB library. 

### ZOSLIB: https://github.com/ibmruntimes/zoslib
* Developed as an extension to the z/OS LE C Runtime Library
* Currently used in Node.js for z/OS product
* Implements a subset of POSIX APIs that are not available in the LE C Runtime Library
* Overrides certain C LE Runtime functions to be more compatible with Linux (activated via macro -DCLIB_OVERRIDES=1
* Implements EBCDIC <-> ASCII file tagging functions, codepage conversion C APIs and more!

z/OS Open Tools zoslibport repo: https://github.com/ZOSOpenTools/zoslibport

**ZOSLIB reference**: [https://ibmruntimes.github.io/zoslib/files.](https://ibmruntimes.github.io/zoslib/files.html)[html](https://ibmruntimes.github.io/zoslib/files.html)

**How to leverage ZOSLIB in z/OS Open Tools?** Will be discussed in Guild meeting 3 and documented in https://zosopentools.github.io/meta/#/

Please let us know if you have any feedback.

**Comment by v1gnesh** [comment]
Hello Igor,

Could you help me understand what this library helps to achieve.
I see the bullet points, but with my limited knowledge on this topic, I'm unable to "get it" :)
What new things will this allow a sysprog to do on z/OS?

The following is mostly unrelated:
I see that [libuv](https://github.com/libuv/libuv) supports SystemZ, since Node uses it.
To keep moving in the direction of supporting performant async on the platform (and enabling zIIP offload of sub/co-routine based languages - Golang's goroutine, Python's asyncio), will [uvloop](https://github.com/magicstack/uvloop) be ported?
Here's a helpful blog post on it - https://magic.io/blog/uvloop-blazing-fast-python-networking/

**Reply by IgorTodorovskiIBM** [reply]
Hi v1gnesh, I'm going to cover this topic in tomorrow's Guild meeting.  But essentially, it helps bridge the gap between the z/OS C LE runtime and glibc.  It also provides APIs for improved use of the z/OS enhanced ASCII support as well as APIs to z/OS MVS services. The idea is that the applications we port would link to ZOSLIB, reducing the effort in modifying the project code to add missing LE functions, etc and have a consistent set of APIs and environment variables to control currently undefined behaviour like how untagged files are read.

Thanks for the uvloop sugestion, I'll relay this to the Python on z/OS lead.

**Reply by v1gnesh** [reply]
Thank you, a further question - considering the role of Metal C (as a replacement interface to use instead of assembler for EXITs), will Metal C also cover some ground here?
That is, like `glibc -> zoslib -> zOS C LE`, can `glibc -> zoslib/Metal C -> zOS C LE` exist, and will it help?

Say there are open source C/C++ libraries or frameworks that are written to extract most out of x86 hardware.
Will having Metal C in the picture here help convert some of the platform-specific code paths (from open source C/C++ libraries/frameworks) to work on zArch?
Ex: To get https://github.com/scylladb/seastar (and subsequently projects/products that use this framework) to work in zOS.

**Reply by IgorTodorovskiIBM** [reply]
Metal C could help, but as the name suggests it supports C only, does not support XPLINK and also is only available with the XLC compiler as of now.  It also bypasses C LE, so it's a good choice if you want to completely replace the z/OS LE with a glibc implementation. Rocket did something similar with the ambitus project. 

We tend to use the inline assembly support when we need to call hlasm macros or instruct the compiler to assemble a sequence of s390 instructions.

**Reply by v1gnesh** [reply]
Another question... what's the difference between [zoslibport](https://github.com/ZOSOpenTools/zoslibport) and [zoslib](https://github.com/ibmruntimes/zoslib)?
Is the former for USS environments, and the latter the actual core for the former?

One more unrelated question (sorry) - what's the reason/requirement for IBM's work in both glibc and LLVM toolchains? Is the glibc work mainly to support zoslib and Metal C (i.e., an LE-free toolchain)?
And if so, will LLVM-based progress (Open C/C++ and beyond) then rely on zoslib for the interaction?

**Reply by IgorTodorovskiIBM** [reply]
Hey, good question. Since our tools have dependencies on each other and since we wanted a generalized CI/CD pipeline, we require that all of our projects adhere to the zopen framework (you can find this in the utils repo).  So essentially, zoslibport is a set of build instructions (defined in the buildenv file) that tell the zopen-build how to build/test/install zoslib.

As far as I know, IBM is not working on a glibc port. IBM did port the LLVM C++ runtime (includes STL and others), but not glibc. The LLVM C++ runtime is used by both the xlclang and ibm-clang compilers.

**Reply by v1gnesh** [reply]
Ah ok, I assumed glibc is ongoing due to how I interpreted 'zoslib is a bridge between LE and glibc'.
Got it now, `xyzport` repos are just projects for holding the build scripts for `xyz`.

Would you be able to share any info on IBM's work for bringing in other LLVM-based languages, to zOS?
